tools/codegen/gen.py

```diff
@@ -201,6 +201,102 @@ def is_cuda_dispatch_key(dk: DispatchKey) -> bool:
+    if not backends or f.manual_kernel_registration:
+        return ''
+
+    is_factory_method = str(f.func.name.name).endswith('_like') or str(f.func.name.name).startswith('new_')
+    tensor_opts = [a for a in cpp_sig.arguments() if isinstance(a.argument, TensorOptionsArguments)]
+    tensor_args = [
+        a.name for a in cpp_sig.arguments()
+        if isinstance(a.argument, SelfArgument) or isinstance(a.argument, Argument) and a.argument.type.is_tensor_like()
+    ]
+    has_backend_select = not is_factory_method and tensor_opts
+
+    if method:
+        tensor_args.append('const_cast<Tensor&>(*this)')
+
+    if f.structured or f.structured_delegate:
+        # structured native functions support CppSignature directly, but it's always not faithful (for now)
+        struct_cpp_sig = CppSignatureGroup.from_native_function(f, method=False, fallback_binding=False).signature
+        name = struct_cpp_sig.name()
+        exprs = translate(
+            cpp_sig.arguments(), struct_cpp_sig.arguments(),
+            method=method, skip_possibly_redundant_memory_format=True)
+        exprs_str = ', '.join(a.expr for a in exprs)
+    else:
+        # translate CppSignature into NativeSignature
+        name = native.name(f.func)
+        exprs = translate(
+            cpp_sig.arguments(), NativeSignature(f.func).arguments(),
+            method=method, skip_possibly_redundant_memory_format=True)
+        exprs_str = ', '.join(a.expr for a in exprs)
+
+    stmts: List[str] = []
+    if not tensor_opts and not tensor_args:
+        if len(f.dispatch) == 1 and DispatchKey.Math in f.dispatch:
+            return f"""\
+    return at::{DispatchKey.Math.lower()}::{name}({exprs_str});
+"""
+        else:
+            return ''
+
+    if not tensor_args:
+        assert len(tensor_opts) == 1
+        # specialized fast pass
+        stmts.append(f"""\
+    DispatchKey _dk = {tensor_opts[0].name}.computeDispatchKey();
+""")
+    else:
+        subexprs: List[str] = []
+        args = ', '.join(tensor_args)
+        subexprs.append(f'c10::detail::multi_dispatch_key_set({args})')
+        for tensor_opt in tensor_opts:
+            subexprs.append(f'DispatchKeySet({tensor_opt.name}.computeDispatchKey())')
+        stmts.append(f"""\
+    DispatchKeySet _dk_set = {' | '.join(subexprs)};
+    DispatchKey _dk = c10::impl::dispatchTypeId(_dk_set, DispatchKeySet::FULL);""")
+
+    stmts.append("""\
+    switch (_dk) {\
+""")
+
+    if not has_backend_select:
+        stmts.append("""\
+    case DispatchKey::BackendSelect:
+        // fallthrough\
+""")
+
+    has_any = False
+    for case_key in backends:
+        for dispatch_key in (case_key, DispatchKey.DefaultBackend, DispatchKey.Math):
+            # FIXME: how do I get dispatch table for function with structured_delegate? Is it correct to
+            # always statically dispatch to the delegate?
```

Comment:
When there's a structured delegate, the dispatch table is automatically generated based on the out variant (https://github.com/pytorch/rfcs/blob/rfc-0005/RFC-0005-structured-kernel-definitions.md#structured-keyword-proposal ; there is no dispatch table for `upsample_nearest1d` because it delegates its dispatch to `upsample_nearest1d_out`). We're still on the hook for generating wrapper functions for all the variants.

I'm not sure if that helps you globally here, still reading.